### ISSUE TITLE
Braille does not work on secure Windows screens while normal copy of NVDA is running

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#2315]
**Reported by kevinchao89 on 2012-05-10 13:40**
When using APH RefreshaBraille, there is no Braille support when on Windows logon screen.

Steps to Reproduce:
1. Ensure NVDA is configured with APH RefreshaBraille 18;
2. WINDOWS+L to lock Windows
3. Read content using Braille display
Expected: read, navigate, and input suing APH RefreshaBraille 18
Actual: Speech-only support/access

**Comment by leonardder** [comment]
I wonder whether this could have been fixed in the new Baum driver, which also applies to this display.

**Comment by jcsteh** [comment]
Not sure what you mean by "new Baum driver". While there have been significant updates, they've all been incremental updates to the original driver that landed in 2010.

Regardless, this isn't fixed. As I noted in https://github.com/nvaccess/nvda/issues/2315#issuecomment-155297694:

> The problem here is that the normal copy of NVDA is using the braille display, so the copy running on the secure screen can't. We need to suspend braille output when switching to secure screens. This is unfortunately much harder than it sounds. :)

This is more of a core issue than a driver specific issue.

**Comment by leonardder** [comment]
[**@leonardder**](https://github.com/leonardder) commented on [12 aug. 2017 11:32 CEST](https://github.com/nvaccess/nvda/issues/2315#issuecomment-321969054 "2017-08-12T09:32:11Z - Replied by Github Reply Comments"):
> I wonder whether this could have been fixed in the new Baum driver, which also applies to this display.

**Comment by leonardder** [comment]
Woops, sorry for going mad there.

[**@jcsteh**](https://github.com/jcsteh) commented on [14 aug. 2017 09:05 CEST](https://github.com/nvaccess/nvda/issues/2315#issuecomment-322114576 "2017-08-14T07:05:19Z - Replied by Github Reply Comments"):
> Regardless, this isn't fixed. As I noted in https://github.com/nvaccess/nvda/issues/2315#issuecomment-155297694:
> 
> > The problem here is that the normal copy of NVDA is using the braille display, so the copy running on the secure screen can't. We need to suspend braille output when switching to secure screens. This is unfortunately much harder than it sounds. :)
> 
> This is more of a core issue than a driver specific issue.

Is this really so hard? We have gainFocus and leaveFocus event for the secure desktop, so couldn't braille just be de-initialized for gainFocus and initialized for leaveFocus?

**Comment by jcsteh** [comment]
[**@leonardder**](https://github.com/leonardder) commented on [14 Aug 2017, 18:27 GMT+10](https://github.com/nvaccess/nvda/issues/2315#issuecomment-322129348 "2017-08-14T08:27:55Z - Replied by Github Reply Comments"):
> Is this really so hard? We have gainFocus and leaveFocus event for the secure desktop, so couldn't braille just be de-initialized for gainFocus and initialized for leaveFocus?

Ah, if only things were as simple as it seems they should be. :) This essentially comes down to timing. What you suggest would work just fine if the secure copy always initialised after the normal copy received the gain focus, finished terminating the braille display driver and Windows released all resources associated with the display. The reverse must also apply when terminating the secure copy. In practice, things don't necessarily happen quite this smoothly. I think the secure copy would need to retry initialisation (perhaps after a short delay) if initialisation failed the first time. The normal copy would likewise need to do this when re-initialising braille after returning from a secure screen.

**Comment by leonardder** [comment]
[**@jcsteh**](https://github.com/jcsteh) commented on [14 aug. 2017 13:20 CEST](https://github.com/nvaccess/nvda/issues/2315#issuecomment-322165249 "2017-08-14T11:20:26Z - Replied by Github Reply Comments"):
> ... Windows released all resources associated with the display...

Talking about freeing resources and access permissions, have you considered opening the HID file handles with FILE_SHARE_READ and FILE_SHARE_WRITE access parameters? Also, how about default permissions on the handles that pyserial creates?

**Comment by jcsteh** [comment]
I think having them open with shared access is risky. Writes from multiple
clients might get interspersed, which could cause all sorts of pain.

**Comment by leonardder** [comment]
Based on NVDA Remote, we could also go for a completely different implementation. When NVDA is running under the current user profile and the secure desktop is activated, create a socket that allows sending speech sequences and braille cell output from the secure copy to the user copy. In that case, speech and braille on secure screens will always match the settings for the local user. This becomes problematic when user switching comes into view or when the user copy of NVDA freezes in such a way that it couldn't longer output speech or braille. However, if the local copy acknowledges every "packet" it receives, the secure screen copy could easily fall back to stand alone speech and braille output as in the current situation when no acknowledgement have been received within a certain period.

**Comment by leonardder** [comment]
It is still present. My advise would be to wait for Python 3 and a possible integration of NVDA Remote into core.

**Comment by leonardder** [comment]
@seanbudd You gave this a milestone of 2023.1, is this something NV Access wants to consider implementing?

I'm currently intending to write a proto type that adds a special braille display driver for this to NVDA that will be active on the secure screen. The secure screen copy of NVDA will send braille to this driver, which will use a named pipe to communicate the braille output to a named pipe server that will be running in the copy of NVDA that's currently active in the user session.
This is also related to https://github.com/nvaccess/nvda/issues/3564#issuecomment-1363206466 in that said driver can be reused for a remote desktop implementation, the underlying connection will be the only thing that differs, but basically it's just writing data to a file handle. The Rust backend I wrote for the remote desktop connection is also based on named pipes, so lots of intentional overlap in scope of the two projects.

**Comment by seanbudd** [comment]
It was initially a serious candidate for 2023.1. After the delays to 2022.4/2022.3.3 we are yet to re-evaluate, but I imagine we won't be completing this in 2023.1. If you do intend to implement a fix for this issue, I suggest doing it separately to the project you describe.

**Comment by leonardder** [comment]
Not sure how #14531 closed this issue, but this is certainly not fixed. #14531 was a step in the right direction though.

**Comment by seanbudd** [comment]
@leonardder "fix #2315" was in the body of the PR description

**Comment by burmancomp** [comment]
@leonardder just doing curiosity not wanting to cause any pressure, referring to what you write here [https://github.com/leonardder/nvdaRd/issues/3](), do you have plans for this issue?

**Comment by leonardder** [comment]
It's already part of the add-on but it is not yet stable enough to be 
ported to NVDA core.

**Comment by leonardder** [comment]
Not very near, but certainly somewhere in the future.

**Comment by SaschaCowley** [comment]
@LeonarddeR any update on this?

**Comment by LeonarddeR** [comment]
This is covered by [RDAccess](https://github.com/leonardder/rdAccess/), though I still classify it as somewhat experimental.

**Comment by LeonarddeR** [comment]
I have recently managed to find the cause of a glitch in [RDAccess](https://github.com/leonardder/rdAccess/) and now the secure desktop support in there is pretty stable.
I might be able to port this to core next summer, however I'm not yet sure about it being the best option. @bramd rightly pointed out privately that even if Braille works on the secure desktop this way, there's still the problem about one instance of NVDA running on user 1 holding the display simply blocks user 2 from using it.
Note that NVDA Remote is part of core now, so we could also go with the approach I suggested in https://github.com/nvaccess/nvda/issues/2315#issuecomment-375002748. There's still the case of the user copy freezing, making the secure desktop unusable.

**Comment by seanbudd** [comment]
@LeonarddeR - do you still plan on doing this?

**Comment by LeonarddeR** [comment]
@seanbudd I could consider doing this, but there's still much unclear about the Implementation of preference (i.e. port the approach used in RD Access or use the NVDA Remote logic).

**Comment by seanbudd** [comment]
Could you share what the approach in RD access looks like, e.g. a link to the code?

**Comment by LeonarddeR** [comment]
The RD Access approach is a bit complex to show. It uses:
- A remote braille display driver and remote speech synth driver
- A custom binary protocol, including pickled data, to transfer data over the wire
- A named pipe to transfer the data from secure desktop to the running user copy

It is probably way easier to use the existing NVDA Remote protocol for this.

**Comment by seanbudd** [comment]
Lets go with the Remote protocol

**Comment by SaschaCowley** [comment]
Uh, what about users who have Remote Access disabled in their user and/or secure config?

**Comment by LeonarddeR** [comment]
That's likely to become problematic indeed, unless we initialize at least a part of the remote client in this case.
Or it is something we need to document. Basically, letting the secure desktop forward its speech and braille to a running copy on the user desktop is a way where a security barrier is crossed anyway. It is also a form of remote control in the strictest sense.

**Comment by SaschaCowley** [comment]
How much investigation has already been conducted into ensuring NVDA releases Braille Displays at secure desktop transitions? Is this a timing issue where the secure desktop copy needs to wait a moment and retry connecting to the braille display?

**Comment by SaschaCowley** [comment]
@LeonarddeR do you plan to work on this? If so, do you need anything from us to continue?

**Comment by LeonarddeR** [comment]
I must do a first round of investigation to see how much time it will cost to implement this, since I have much less free time starting from september. It also depends on how much priority this has for NV Access.
I'm also not sure how to run local NVDA on the secure desktop. Do I need signed builds for that? I recall that's not the case.

**Comment by SaschaCowley** [comment]
I think you need a signed build to install to the secure desktop, but a self-signed build should work fine as long as you install your self-signing certificate as a trusted root certificate (remember to remove it when you're done).
